### PR TITLE
fix(SelectionBuilder): Add SelectionBuilder

### PR DESCRIPTION
**Pull request by jourdain**
Add SelectionBuilder to improve selection handling across widgets

### CODE REVIEW

**Comment by jourdain** [comment]
@vibraphone @aronhelser @scottwittenburg

File: src/Common/Misc/SelectionBuilder/index.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+  });
+
+  return selection;
+}
+
+// ----------------------------------------------------------------------------
+
+export function rule(type = 'multi', terms = [], roles = []) {
+  generation++;
+  // FIXME ?? deepClone ??
```
Comment (aronhelser) [inline src/Common/Misc/SelectionBuilder/index.js]: deepClone? Anything specific you are worried about?

File: src/Common/Misc/SelectionBuilder/index.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    }
+    if (Array.isArray(clonedObj[name])) {
+      clonedObj[name] = clonedObj[name].map(i => i);
+    }
+  });
+  return clonedObj;
+}
+
+const endpointToRuleOperator = {
+  o: '<',
```
Comment (aronhelser) [inline src/Common/Misc/SelectionBuilder/index.js]: Need a comment indicating 'open' and 'closed'

**Comment by jourdain** [comment]
Sorry @aronhelser, my mouse couldn't resist to the green button.
I just notice your comment after it was to late.

I'm not worried that much, and that class will need some add-on later on, so we can add those missing comments and code update at that point.

File: src/Common/Misc/SelectionBuilder/index.js (hunk 1 above)
Comment (jourdain) [inline src/Common/Misc/SelectionBuilder/index.js]: not really ;-)

**Comment by aronhelser** [comment]
Yep, fine with me - I need to build on it. 

But there are almost no comments - we need to know what 3L and 5C mean if we haven't just read David's document :)

**Pull request merged**